File: backend/__SLUG___node/src/logger/logger.ts
import pino, { type Logger } from "pino";
import { AsyncLocalStorage } from "node:async_hooks";

interface RequestContext {
  logger: Logger;
  traceId: string;
  spanId: string;
}

const als = new AsyncLocalStorage<RequestContext>();

const gcpLevelMap: Record<string, string> = {
  trace: "DEBUG",
  debug: "DEBUG",
  info: "INFO",
  warn: "WARNING",
  error: "ERROR",
  fatal: "CRITICAL",
};

/** Maps pino levels to GCP Cloud Logging severity strings, defaulting to DEFAULT for unknown levels. */
export function gcpLevelEncoder(level: string): string {
  return gcpLevelMap[level] ?? "DEFAULT";
}

let rootLogger: Logger = pino({ level: "info" });

export function initLogger(): void {
  const level = process.env.LOG_LEVEL === "debug" ? "debug" : "info";

  if (process.env.LOG_PRETTY === "true") {
    rootLogger = pino({
      level,
      transport: { target: "pino-pretty" },
    });
    return;
  }

  rootLogger = pino({
    level,
    messageKey: "message",
    timestamp: pino.stdTimeFunctions.isoTime,
    formatters: {
      level(label) {
        return { severity: gcpLevelEncoder(label) };
      },
    },
  });
}

export function withRequestContext<T>(ctx: RequestContext, fn: () => T): T {
  return als.run(ctx, fn);
}

export function loggerFromContext(): Logger {
  return als.getStore()?.logger ?? rootLogger;
}

export function traceIdFromContext(): string {
  return als.getStore()?.traceId ?? "";
}

export function spanIdFromContext(): string {
  return als.getStore()?.spanId ?? "";
}

export function getRootLogger(): Logger {
  return rootLogger;
}

File: backend/__SLUG__/src/logger/logger.go
package logger

import (
	"context"
	"os"

	"go.uber.org/zap"
	"go.uber.org/zap/zapcore"
)

type contextKey int

const (
	loggerKey  contextKey = iota
	traceIDKey contextKey = iota
	spanIDKey  contextKey = iota
)

func InitLogger() {
	var core zapcore.Core

	level := zapcore.InfoLevel
	if os.Getenv("LOG_LEVEL") == "debug" {
		level = zapcore.DebugLevel
	}

	if os.Getenv("LOG_PRETTY") == "true" {
		encCfg := zap.NewDevelopmentEncoderConfig()
		encCfg.EncodeLevel = zapcore.CapitalColorLevelEncoder
		core = zapcore.NewCore(
			zapcore.NewConsoleEncoder(encCfg),
			zapcore.AddSync(os.Stdout),
			level,
		)
	} else {
		encCfg := zap.NewProductionEncoderConfig()
		encCfg.MessageKey = "message"
		encCfg.LevelKey = "severity"
		encCfg.TimeKey = "time"
		encCfg.EncodeTime = zapcore.RFC3339NanoTimeEncoder
		encCfg.EncodeLevel = gcpLevelEncoder
		core = zapcore.NewCore(
			zapcore.NewJSONEncoder(encCfg),
			zapcore.AddSync(os.Stdout),
			level,
		)
	}

	zap.ReplaceGlobals(zap.New(core))
}

// gcpLevelEncoder maps zap levels to GCP Cloud Logging severity strings.
func gcpLevelEncoder(l zapcore.Level, enc zapcore.PrimitiveArrayEncoder) {
	switch l {
	case zapcore.DebugLevel:
		enc.AppendString("DEBUG")
	case zapcore.InfoLevel:
		enc.AppendString("INFO")
	case zapcore.WarnLevel:
		enc.AppendString("WARNING")
	case zapcore.ErrorLevel:
		enc.AppendString("ERROR")
	case zapcore.DPanicLevel, zapcore.PanicLevel, zapcore.FatalLevel:
		enc.AppendString("CRITICAL")
	default:
		enc.AppendString("DEFAULT")
	}
}

// WithLogger stores a scoped logger in the context.
func WithLogger(ctx context.Context, l *zap.Logger) context.Context {
	return context.WithValue(ctx, loggerKey, l)
}

// LoggerFromContext retrieves the scoped logger from context, falling back to the global logger.
func LoggerFromContext(ctx context.Context) *zap.Logger {
	if l, ok := ctx.Value(loggerKey).(*zap.Logger); ok && l != nil {
		return l
	}
	return zap.L()
}

// WithTraceID stores trace and span IDs in the context.
func WithTraceID(ctx context.Context, traceID, spanID string) context.Context {
	ctx = context.WithValue(ctx, traceIDKey, traceID)
	ctx = context.WithValue(ctx, spanIDKey, spanID)
	return ctx
}

// TraceIDFromContext retrieves the trace ID from context.
func TraceIDFromContext(ctx context.Context) string {
	if id, ok := ctx.Value(traceIDKey).(string); ok {
		return id
	}
	return ""
}

// SpanIDFromContext retrieves the span ID from context.
func SpanIDFromContext(ctx context.Context) string {
	if id, ok := ctx.Value(spanIDKey).(string); ok {
		return id
	}
	return ""
}
